exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,709 @@
+                                                .map_err(into_failure)?;
+                                let message = Message::new(payload, message);
+                                Ok($protocol_name::$class($class::$type(message)))
+                            }),+
+                            _ => bail!("Not found message with this type {}", message.message_type())
+                        }
+                    ),+
+                    _ => bail!("Not found message with this class {}", message.message_class())
+                }
+            }
+
+            /// Returns reference to inner `SignedMessage`.
+            pub fn signed_message(&self) -> &SignedMessage {
+                match *self {
+                    $(
+                        $protocol_name::$class(ref c) => {
+                            match *c {
+                                $(
+                                    $class::$type(ref t) => {
+                                        t.signed_message()
+                                    }
+                                ),+
+                            }
+                        }
+                    ),+
+                }
+            }
+        }
+    };
+}
+
+impl_protocol!{
+    /// Composition of every exonum protocol messages.
+    /// This messages used in network p2p communications.
+    SignedMessage => Protocol {
+        /// Exonum basic node messages.
+        0 => Service {
+            /// `RawTransaction` representation.
+            RawTransaction = 0,
+            /// Handshake to other node.
+            Connect = 1,
+            /// `Status` information of other node.
+            Status = 2,
+        },
+        /// Exonum consensus specific node messages.
+        1 => Consensus {
+            /// Consensus `Precommit` message.
+            Precommit = 0,
+            /// Consensus `Propose` message.
+            Propose = 1,
+            /// Consensus `Prevote` message.
+            Prevote = 2,
+        },
+        /// Exonum node responses.
+        2 => Responses {
+            /// Information about transactions, that sent as response to `TransactionsRequest`.
+            TransactionsResponse = 0,
+            /// Information about block, that sent as response to `BlockRequest`.
+            BlockResponse = 1,
+        },
+        /// Exonum node requests.
+        3 => Requests {
+
+            /// Request of some propose which hash is known.
+            ProposeRequest = 0,
+            /// Request of unknown transactions.
+            TransactionsRequest = 1,
+            /// Request of prevotes for some propose.
+            PrevotesRequest = 2,
+            /// Request of peer exchange.
+            PeersRequest = 3,
+            /// Request of some future block.
+            BlockRequest = 4,
+        },
+
+    }
+}
+
+impl Protocol {
+    /// Creates new protocol message.
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
```

Comment:
(First of all, correct grammar would be "this method can panic on serialization failure".)

What exactly are the conditions for a panic? Is the user expected to provide some 'correct' arguments to avoid the panic, or is this just a warning that serialization may fail for example due to memory allocation issues?

In the former case I would describe the specific expectations (e.g., that the secret and public keys must match). Otherwise I think we can omit the panic section if avoiding the panic is not something under caller's control.

Reply:
@ilammy 
>What exactly are the conditions for a panic? Is the user expected to provide some 'correct' arguments to avoid the panic, or is this just a warning that serialization may fail for example due to memory allocation issues?

Both, based on serialisation implementation. 
For example our `exonum_encoding` never return error. It will panic on OOM internaly.
Serde json for example can return error based on internal conditions.

Probably service creator shouldn't ever use this methods. But we should notice core developers who probably will use it, that they should provide unfailable `BinaryForm::serialize` implementation.